java/src/main/java/cucumber/runtime/java/ObjectFactoryLoader.java

```diff
@@ -53,6 +55,46 @@ public static ObjectFactory loadObjectFactory(ClassFinder classFinder, String ob
         return objectFactory;
     }
 
+    private static io.cucumber.core.backend.ObjectFactory loadSingleObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactories = loadObjectFactories(reflections, loader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactories.isEmpty()) {
+            throw new NoInstancesException(io.cucumber.core.backend.ObjectFactory.class);
+        } else if (objectFactories.size() > 1) {
+            throw new TooManyInstancesException(objectFactories.values());
+        } else {
+            result = objectFactories.values().iterator().next();
+        }
+        return result;
+    }
+
+    private static Map<String, io.cucumber.core.backend.ObjectFactory> loadObjectFactories(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader) {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> knowObjectFactories = new HashMap<>();
+        final List<URI> packages = asList(URI.create("classpath:cucumber/runtime"));
+        loadObjectFactories(reflections.instantiateSubclasses(io.cucumber.core.backend.ObjectFactory.class, packages, new Class[0], new Object[0]), knowObjectFactories);
+        loadObjectFactories(loader, knowObjectFactories);
+        return knowObjectFactories;
+    }
+
+    private static io.cucumber.core.backend.ObjectFactory loadSelectedObjectFactory(final Reflections reflections, final ServiceLoader<io.cucumber.core.backend.ObjectFactory> loader, final ClassFinder classFinder, final String objectFactoryClassName) throws ClassNotFoundException {
+        final Map<String, io.cucumber.core.backend.ObjectFactory> objectFactoriesByServiceLoader = new HashMap<>();
+        loadObjectFactories(loader, objectFactoriesByServiceLoader);
+        final io.cucumber.core.backend.ObjectFactory result;
+        if (objectFactoriesByServiceLoader.containsKey(objectFactoryClassName)) {
+            result = objectFactoriesByServiceLoader.get(objectFactoryClassName);
+        } else {
+            final Class<io.cucumber.core.backend.ObjectFactory> objectFactoryClass = (Class<io.cucumber.core.backend.ObjectFactory>) classFinder.loadClass(objectFactoryClassName);
+            result = reflections.newInstance(new Class[0], new Object[0], objectFactoryClass);
+        }
+        return result;
+    }
+
+    private static void loadObjectFactories(final Iterable<? extends io.cucumber.core.backend.ObjectFactory> objectFactories, final Map<String, io.cucumber.core.backend.ObjectFactory> knowObjectFactories) {
+        for (io.cucumber.core.backend.ObjectFactory objectFactory : objectFactories) {
```

Comment:
This will actually instantiate all object factories. This is never desirable.

Reply:
Ah. I see you copied this from the original. I should fix it there too.

Reply:
Yeah but why is this a issue? If you have more than one at your path cucumber ignore them an than run all scenarios with the default one with take more time than create unnecessary instances? And if you look at the code in reflector it does the same?

Reply:
It avoids potential side effects. Means there is one less thing to take into account when debugging. Just because it was done in the past, doesn't mean it's good idea to continue it.